src/devices/src/virtio/block/device.rs

```diff
@@ -332,886 +332,594 @@ impl VirtioDevice for Block {
+        {
+            let features: u64 = (1u64 << VIRTIO_BLK_F_RO)
+                | (1u64 << VIRTIO_F_VERSION_1)
+                | (1u64 << VIRTIO_BLK_F_FLUSH);
+
+            assert_eq!(block.avail_features_by_page(0), features as u32);
+            assert_eq!(block.avail_features_by_page(1), (features >> 32) as u32);
+            for i in 2..10 {
+                assert_eq!(block.avail_features_by_page(i), 0u32);
+            }
+
+            for i in 0..10 {
+                block.ack_features_by_page(i, u32::MAX);
+            }
+            assert_eq!(block.acked_features, features);
+        }
+
+        // Test `write_config()`.
+        {
+            let new_config: [u8; 8] = [0x00, 0x50, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00];
+            block.write_config(0, &new_config);
+            let mut new_config_read = [0u8; 8];
+            block.read_config(0, &mut new_config_read);
+            assert_eq!(new_config, new_config_read);
+            // Invalid write.
+            check_metric_after_block!(
+                &METRICS.block.cfg_fails,
+                1,
+                block.write_config(5, &new_config)
+            );
+            // Make sure nothing got written.
+            new_config_read = [0u8; 8];
+            block.read_config(0, &mut new_config_read);
+            assert_eq!(new_config, new_config_read);
+        }
+    }
+
+    #[test]
+    #[allow(clippy::cognitive_complexity)]
+    fn test_events() {
+        // Create a default Block with a virtqueue that we can write to.
+        let m = GuestMemory::new(&[(GuestAddress(0), 0x10000)]).unwrap();
+        let vq = VirtQueue::new(GuestAddress(0), &m, 16);
+        assert!(vq.end().0 < 0x1000);
+
+        let mut block = default_block();
+        block.queues = vec![vq.create_queue()];
+        block.mem = m.clone();
+
+        let blk_metadata = block.disk_image.metadata();
+
+        for i in 0..3 {
+            vq.avail.ring[i].set(i as u16);
+            vq.dtable[i].set(
+                (0x1000 * (i + 1)) as u64,
+                0x1000,
+                VIRTQ_DESC_F_NEXT,
+                (i + 1) as u16,
+            );
+        }
+
+        vq.dtable[1]
+            .flags
+            .set(VIRTQ_DESC_F_NEXT | VIRTQ_DESC_F_WRITE);
+        vq.dtable[2].flags.set(VIRTQ_DESC_F_WRITE);
+        vq.avail.idx.set(1);
+
+        // dtable[1] is the data descriptor.
+        let data_addr = GuestAddress(vq.dtable[1].addr.get());
+        // dtable[2] is the status descriptor.
+        let status_addr = GuestAddress(vq.dtable[2].addr.get());
+
+        // Test invalid request.
+        {
+            // Request is invalid because the first descriptor is write-only.
+            vq.dtable[0]
+                .flags
+                .set(VIRTQ_DESC_F_NEXT | VIRTQ_DESC_F_WRITE);
+            m.write_obj_at_addr::<u32>(VIRTIO_BLK_T_IN, GuestAddress(0x1000))
+                .unwrap();
+
+            invoke_handler_for_queue_event(&mut block);
```

Comment:
I would prefer an actual `handle_read` call instead of a function which invokes the handling function, because it will resemble how the handling is done in the production code.

Reply:
I replaced all calls to `process_*` with calls to `handle_read`.